signer/src/network.rs

```diff
@@ -14,89 +14,74 @@ use crate::message;
 
 pub type Msg = ecdsa::Signed<message::SignerMessage>;
 
+/// Represents the interaction point between signers and the signer network,
+/// allowing signers to exchange messages with each other.
 pub trait MessageTransfer {
-    type Error;
+    type Error: std::error::Error;
     /// Send `msg` to all other signers
     fn broadcast(&mut self, msg: Msg) -> impl Future<Output = Result<(), Self::Error>> + Send;
     /// Receive a message from the network
     fn receive(&mut self) -> impl Future<Output = Result<Msg, Self::Error>> + Send;
 }
 
 #[cfg(test)]
-mod tests {
-    use super::*;
-
-    use bitcoin::hashes::Hash;
-    use p256k1::scalar;
+pub async fn assert_clients_should_be_able_to_exchange_messages<
+    C: MessageTransfer + Send + 'static,
+>(
+    mut client_1: C,
+    mut client_2: C,
+) {
     use rand::SeedableRng;
+    let mut rng = rand::rngs::StdRng::seed_from_u64(1337);
+    let number_of_messages = 32;
+
+    let client_1_messages: Vec<_> = (0..number_of_messages)
+        .map(|_| Msg::random(&mut rng))
+        .collect();
+    let client_2_messages: Vec<_> = (0..number_of_messages)
+        .map(|_| Msg::random(&mut rng))
+        .collect();
+
+    let client_1_expected_received_messages = client_2_messages.clone();
+    let client_2_expected_received_messages = client_1_messages.clone();
+
```

Comment:
Can you add a comment sentence or two describing what's going on here? Something like:

```rust
// spawn two processes that will send a single message to eachother and read the message
// that the other one sent. This will only verify that the clients can talk to eachother, but
// does not validate any part of the message handling.
```

Reply:
Definitely, will do :saluting_face:

Reply:
Okay working on this, I realize that this makes a good docstring for the test, so I will add it there with some corrections.

Reply:
I added a docstring, and broke out the shared logic so I hope the test helper is more clear and readable now. Let me know what you think of the updated structure.